meteor/sofie-server/business-logic/services/superfly-timeline-builder.ts

```diff
@@ -0,0 +1,435 @@
+				return {
+					start: `#${parentGroup.id}.end - ${part.outTransition.keepAliveDuration}${postRollDurationPostFix}`,
+				}
+			}
+			case TransitionType.NO_TRANSITION: {
+				// TODO: Check for dynamically inserted and if the Piece.duration is a number (which it should be).
+				// TODO: If it is, set enable.start += partCalculatedTimings.toPartDelay-
+				// TODO: This is something Core does. Might just be used for AdLibs, if used at all?
+
+				// TODO: This hurts...
+				const duration: string | number =
+					partCalculatedTimings.postRollDuration && !piece.duration
+						? `#${parentGroup.id} - ${partCalculatedTimings.postRollDuration}`
+						: piece.duration
+
+				return {
+					start: piece.start,
+					duration: duration,
+				}
+			}
+			default: {
+				throw new UnsupportedOperation(
+					`Unknown TransitionType "${piece.transitionType}" for Piece: ${piece.id}`
+				)
+			}
+		}
+	}
+
+	private createTimelineObjectControl(
+		parentGroup: TimelineObjectGroup,
+		piece: Piece,
+		pieceEnable: TimelineEnable
+	): TimelineObjectControl {
+		return {
+			id: `${parentGroup.id}${PIECE_CONTROL_INFIX}${piece.id}`,
+			enable: pieceEnable,
+			layer: piece.layer,
+			priority: MEDIUM_PRIORITY,
+			content: {
+				type: TimelineObjectType.CONTROL,
+				deviceType: DeviceType.ABSTRACT,
+			},
+			isGroup: false,
+			inGroup: parentGroup.id,
+		}
+	}
+
+	private createTimelineObjectGroupForPiece(
+		parentGroup: TimelineObjectGroup,
+		piece: Piece,
+		controlForPiece: TimelineObjectControl
+	): TimelineObjectGroup {
+		return {
+			id: `${parentGroup.id}${PIECE_GROUP_INFIX}${piece.id}`,
+			content: {
+				type: TimelineObjectType.GROUP,
+				deviceType: DeviceType.ABSTRACT,
+			},
+			isGroup: true,
+			inGroup: parentGroup.id,
+			children: [],
+			enable: {
+				start: `#${controlForPiece.id}.start - ${piece.preRollDuration ?? 0}`,
+				end: `#${controlForPiece.id}.end - ${piece.postRollDuration ?? 0}`,
+			},
+			layer: '',
+		}
+	}
+
+	private shouldPieceHavePreRollGroup(controlForPiece: TimelineObjectControl, piece: Piece): boolean {
+		// TODO: If we can't get rid of start="now" then we need to check for "now" here.
+		return (controlForPiece.enable as TimelineEnable).start === 0 && piece.preRollDuration > 0
+	}
+
+	private createPreRollGroupForControl(
+		controlObject: TimelineObjectControl,
+		parentGroup: TimelineObjectGroup
+	): TimelineObjectControl {
+		return {
+			id: `${PIECE_PRE_ROLL_PREFIX}${controlObject.id}`,
+			enable: {
+				start: `#${parentGroup.id}.start`,
+			},
+			layer: '',
+			content: {
+				type: TimelineObjectType.CONTROL,
+				deviceType: DeviceType.ABSTRACT,
+			},
+		}
+	}
+
+	private mapToTimelineObjectForPieceGroup(
+		timelineObject: TimelineObject,
+		childGroupForPiece: TimelineObjectGroup,
+		piece: Piece
+	): TimelineObject {
+		// TODO: Core checks "HoldMode" on the timelineObject - ignore for now
+		const timelineObjectCopy: TimelineObject = JSON.parse(JSON.stringify(timelineObject))
```

Comment:
I like the idea about extracting it.

Utility/helper classes are useally a bad practice since the names `utility` and `helper` doesn't indicate a single responsibility, they indicate quite the opposite.

I suggest extracting it into an abstraction called something like `interface ObjectCloneService` and the implementation could be `JsonObjectCloneService`.

If you have naming suggestions, I'm all ears :)

Reply:
I realize you could argue the same for `service` and it's the prefix the gives it the responsibility. 
The important part is there is _something_ to indicate the specific responsibility of the class :)

Reply:
Updated.
If you have any other suggestions for the names, I am all ears. I am not the biggest fan of them myselves.